resources/doxygen/PAC52XX_drivers/man/man3/PAC5XXX_Driver_System.3
.TH "PAC5XXX_Driver_System" 3 "Wed Sep 23 2020" "Version 4.1.2" "FOC-4.1.2-PAC52XX" \" -*- nroff -*-
.ad l
.nh
.SH NAME
PAC5XXX_Driver_System \- PAC5XXX System and Clock Controller Driver Functions
.SH SYNOPSIS
.br
.PP
.SS "Macros"

.in +1c
.ti -1c
.RI "#define \fBRAMFUNC_SYSTEM\fP"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "\fBRAMFUNC_SYSTEM\fP void \fBpac5xxx_sys_ccs_config\fP (\fBCCSCTL_CLKIN_Type\fP clock_select, \fBCCSCTL_ACLKDIV_Type\fP aclk_div, \fBCCSCTL_HCLKDIV_Type\fP fclk_div)"
.br
.RI "This function configures the PAC5XXX clock control system\&. "
.ti -1c
.RI "\fBRAMFUNC_SYSTEM\fP void \fBpac5xxx_sys_pll_config_dividers\fP (uint8_t input_div, uint16_t feedback_div, uint8_t output_div)"
.br
.RI "This function configures the PAC5XXX PLL dividers\&. "
.ti -1c
.RI "\fBRAMFUNC_SYSTEM\fP void \fBpac5xxx_sys_pll_enable\fP (int delay)"
.br
.RI "This function enables the PAC5XXX PLL\&. "
.ti -1c
.RI "\fBRAMFUNC_SYSTEM\fP void \fBpac5xxx_sys_pll_config_100mhz\fP (\fBCCSCTL_HCLKDIV_Type\fP fclk_div, \fBCCSCTL_ACLKDIV_Type\fP aclk_div)"
.br
.RI "This function configures and enables the PLL for a 100MHz output\&. "
.ti -1c
.RI "\fBRAMFUNC_SYSTEM\fP void \fBpac5xxx_sys_pll_config\fP (int frequency_mhz)"
.br
.RI "This function configures and enables the PLL with the given frequency output\&. "
.ti -1c
.RI "void \fBpac_recovery\fP (void)"
.br
.ti -1c
.RI "void \fBpac_delay_asm\fP (uint32_t counts)"
.br
.in -1c
.SH "Detailed Description"
.PP 

.SH "Macro Definition Documentation"
.PP 
.SS "#define RAMFUNC_SYSTEM"

.PP
Definition at line 32 of file pac5xxx_driver_system\&.h\&.
.SH "Function Documentation"
.PP 
.SS "\fBRAMFUNC_SYSTEM\fP void pac5xxx_sys_ccs_config (\fBCCSCTL_CLKIN_Type\fP clock_select, \fBCCSCTL_ACLKDIV_Type\fP aclk_div, \fBCCSCTL_HCLKDIV_Type\fP fclk_div)"

.PP
This function configures the PAC5XXX clock control system\&. 
.PP
\fBParameters\fP
.RS 4
\fIclock_select\fP The input clock select for FRCLK 
.br
\fIaclk_div\fP The ACLK divider 
.br
\fIfclk_div\fP The FCLK divider 
.RE
.PP
\fBReturns\fP
.RS 4
none 
.RE
.PP

.PP
Definition at line 20 of file pac5xxx_driver_system_ccs_config\&.c\&.
.PP
.nf
21 {
22     PAC5XXX_SYSCTL->CCSCTL\&.CLKIN = clock_select;
23     PAC5XXX_SYSCTL->CCSCTL\&.ACLKDIV = aclk_div;
24     PAC5XXX_SYSCTL->CCSCTL\&.HCLKDIV = fclk_div;
25 }
.fi
.PP
References PAC5XXX_SYSCTL\&.
.PP
Referenced by init_system()\&.
.SS "\fBRAMFUNC_SYSTEM\fP void pac5xxx_sys_pll_config (int frequency_mhz)"

.PP
This function configures and enables the PLL with the given frequency output\&. 
.PP
\fBParameters\fP
.RS 4
\fIfrequency_mhz\fP Output frequency in MHz
.RE
.PP
\fBNote\fP
.RS 4
This function configures and enables the PLL from the internal reference (CLKREF)\&. After the 500us wait time (for the PLL to lock), the FCLK output is switched from CLKREF to the PLL output\&. 
.RE
.PP

.PP
Definition at line 21 of file pac5xxx_driver_system_pll_config_mhz\&.c\&.
.PP
.nf
22 {
23     // Switch to the CLKREF input from internal oscillator, Configure FCLK /2, ACLK /1
24     PAC5XXX_SYSCTL->CCSCTL\&.CLKIN = CCSCTL_CLKIN_CLKREF;
25 
26     // Turn off internal oscillator, we don't need it anymore
27     pac5xxx_sys_osc_disable();
28 
29     // Enable PLL (carefully)
30 
31     //
32     // PLLOUT = PLLIN * (PLLFD / (PLLID * PLLOD * 2))
33     // Constraints:
34     //   - PLLFD: 2-514
35     //   - PLLID: 2-34
36     //   - PLLOD: 0-16
37     //
38     // Configure PLL dividers: IN /2 [min], FDBK 100, OUT /1 [min]
39     // should be PLLOUT at 100MHz: 4MHz * (100 / (2 * 100 * 2))
40     pac5xxx_sys_pll_config_dividers(PLLCTL_INPUT_DIV(2), PLLCTL_FDBK_DIV(frequency_mhz), PLLCTL_OUTPUT_DIV(1));
41 
42     // SysTick is configured to use the reference clock (FRCLK)
43     // The STCLK is CLKREF / 3, or 4MHz / 3 = 1\&.333MHz or 0\&.75us/tick\&. So to get a delay of 500us,
44     // you need a count value of 500 / 0\&.75 = 666
45     pac5xxx_sys_pll_enable(666);
46 
47     // Now we can switch to the PLLCLK for the ACLK and FCLK divider input
48     pac5xxx_sys_ccs_pll_select();
49 }
.fi
.PP
References CCSCTL_CLKIN_CLKREF, and PAC5XXX_SYSCTL\&.
.PP
Referenced by init_system()\&.
.SS "\fBRAMFUNC_SYSTEM\fP void pac5xxx_sys_pll_config_100mhz (\fBCCSCTL_HCLKDIV_Type\fP fclk_div, \fBCCSCTL_ACLKDIV_Type\fP aclk_div)"

.PP
This function configures and enables the PLL for a 100MHz output\&. 
.PP
\fBParameters\fP
.RS 4
\fIfclk_div\fP HCLK output divider after PLL stage 
.br
\fIaclk_div\fP ACLK output divider after PLL stage
.RE
.PP
\fBNote\fP
.RS 4
This function configures and enables the PLL from the internal reference (CLKREF), with the given HCLK and FCLK dividers\&. After the 500us wait time (for the PLL to lock), the FCLK output is switched from CLKREF to the PLL output\&. 
.RE
.PP

.SS "\fBRAMFUNC_SYSTEM\fP void pac5xxx_sys_pll_config_dividers (uint8_t input_div, uint16_t feedback_div, uint8_t output_div)"

.PP
This function configures the PAC5XXX PLL dividers\&. 
.PP
\fBParameters\fP
.RS 4
\fIinput_div\fP Input clock divider 
.br
\fIfeedback_div\fP Feedback divider 
.br
\fIoutput_div\fP Output divider 
.RE
.PP
\fBReturns\fP
.RS 4
none 
.RE
.PP

.PP
Definition at line 20 of file pac5xxx_driver_system_pll_config\&.c\&.
.PP
.nf
21 {
22     PAC5XXX_SYSCTL->PLLCTL\&.PLLINDIV = input_div;
23     PAC5XXX_SYSCTL->PLLCTL\&.PLLFBDIV = feedback_div;
24     PAC5XXX_SYSCTL->PLLCTL\&.PLLOUTDIV = output_div;
25 }
.fi
.PP
References PAC5XXX_SYSCTL\&.
.SS "\fBRAMFUNC_SYSTEM\fP void pac5xxx_sys_pll_enable (int delay)"

.PP
This function enables the PAC5XXX PLL\&. 
.PP
\fBParameters\fP
.RS 4
\fIdelay\fP The number of SysTick clocks to delay 
.RE
.PP
\fBReturns\fP
.RS 4
none
.RE
.PP
\fBNote\fP
.RS 4
The user should make sure that PLL bypass is active, by calling the function: \fIpac5xxx_sys_pll_bypass_enable()\fP before calling this function\&.
.PP
There is a mandatory 500us delay after the PLL is enabled before it is stable and ready to use\&. This function will enforce the passed in delay after it enables the PLL before it returns control to the user\&. At that point, the user may remove the PLL bypass by calling \fIpac5xxx_sys_pll_bypass_disable()\fP\&. This time delay is enforced by polling the SysTick timer\&. 
.RE
.PP

.PP
Definition at line 27 of file pac5xxx_driver_system_pll_config\&.c\&.
.PP
.nf
28 {
29   
30   // Configure SysTick the number of SysTick clock ticks passed into this function\&.
31   SysTick->LOAD = delay;
32   SysTick->VAL = 0;
33     
34   // Enable PLL
35   PAC5XXX_SYSCTL->PLLCTL\&.PLLEN = 1;
36   
37   // Enable SysTick timer countdown and set the clock to reference clock
38   SysTick->CTRL |= SysTick_CTRL_ENABLE_Msk;
39   
40   // Wait until it reaches 0 and then return
41   while ((SysTick->CTRL & SysTick_CTRL_COUNTFLAG_Msk) == 0);
42   
43   
44 }
.fi
.PP
References PAC5XXX_SYSCTL, SysTick, SysTick_CTRL_COUNTFLAG_Msk, and SysTick_CTRL_ENABLE_Msk\&.
.SS "void pac_delay_asm (uint32_t counts)"

.PP
Referenced by pac_recovery()\&.
.SS "void pac_recovery (void)"

.PP
Definition at line 42 of file pac5xxx_driver_system_ccs_config\&.c\&.
.PP
.nf
43 {
44     volatile uint32_t i;
45     uint32_t perform_delay = 1;
46 
47     // PE2 is already an input, just need to Enable Pull Up
48     PAC5XXX_GPIOE->PU\&.P2 = 1;                         // Enable Pull Up
49 
50     // If PE2 is high a delay is not needed
51     // Must check over the length of a full UART packet at 115kbps (78uS) in case UART is talking and making PE2 low for short durations
52     for (i=0; i < 12; i++)  // Value of 12 = at least 140 uS, which is enough for 1 UART packet
53     {
54         // If PE2 is high at any time, then a delay is not needed
55         if (PAC5XXX_GPIOE->IN\&.P2 == 1)
56         {
57             perform_delay = 0;
58         }
59     }
60 
61     // If PE2 was low, delay required, add 6 second delay
62     if(perform_delay)
63     {
64         pac_delay_asm(1400000);
65     }
66 
67     // Restore PE2 to reset state
68     PAC5XXX_GPIOE->PU\&.P2 = 0;                         // Disable Pull Up
69 }
.fi
.PP
References PAC5XXX_GPIOE, and pac_delay_asm()\&.
.SH "Author"
.PP 
Generated automatically by Doxygen for FOC-4\&.1\&.2-PAC52XX from the source code\&.


pac55xx/pac55xx_sdk/driver/pac5xxx_driver_system.c
/**************************************************************************//**
 * @file     pac5xxx_driver_system.c
 * @brief    Firmware driver for the PAC5XXX system and clock controllers
 *
 * @note
 * Copyright (C) 2015-2020, Qorvo, Inc.
 *
 * THIS SOFTWARE IS SUBJECT TO A SOURCE CODE LICENSE AGREEMENT WHICH PROVIDES,
 * AMONG OTHER THINGS:  (i) THAT IT CAN BE USED ONLY TO ADAPT THE LICENSEE'S
 * APPLICATION TO PAC PROCESSORS SUPPLIED BY QORVO, INC.;
 * (ii) THAT  IT IS PROVIDED "AS IS" WITHOUT WARRANTY;  (iii) THAT
 * QORVO, INC. IS NOT LIABLE FOR ANY INDIRECT DAMAGES OR FOR DIRECT
 * DAMAGES EXCEEDING US$1,500;  AND (iv) THAT IT CAN BE DISCLOSED TO AND USED
 * ONLY BY CERTAIN AUTHORIZED PERSONS.
 *
 ******************************************************************************/

#include "PAC5XXX_driver_system.h"


RAMFUNC_SYSTEM void pac5xxx_sys_pll_config_enable(uint32_t pllindiv, uint32_t pllfbdiv, uint32_t plloutdiv)
{
    //  Configure PLL control reg and enable PLL
    PAC55XX_SCC->CCSPLLCTL.PLLEN = 0;
    PAC55XX_SCC->CCSPLLCTL.PLLFBDIV = pllfbdiv;
    PAC55XX_SCC->CCSPLLCTL.PLLOUTDIV = plloutdiv;
    PAC55XX_SCC->CCSPLLCTL.PLLINDIV = pllindiv;
    PAC55XX_SCC->CCSPLLCTL.PLLEN = 1;
    // Wait for Lock
    while(PAC55XX_SCC->CCSPLLCTL.PLLLOCK == 0);    
}


//=====================================================================================
// pac_recovery() provides a mechanism to recover a misconfigured PAC device
// Ground PE3 at power up to add a 6 second delay so that the Flash can be erased
//=====================================================================================
// NOTE:  #define PAC_RECOVERY should be commented out for a production system
// When PAC_RECOVERY is defined, then a 6 Second delay will be added if PE3 is connected to ground at PAC power up
// The 6 second delay will give the user time to erase the device using SWD, which will eliminate the code that is misconfiguring the device
// On PAC55xx EVKs, PE3 is normally the GUI UART PAC RX input pin (Host TX); a delay will not normally be inserted even with GUI running
#define PAC_RECOVERY    // Define to enable PAC recovery mechanism

#ifdef PAC_RECOVERY

#warning PAC_RECOVERY enabled; Ground PE3 to add 6 second boot delay to allow recovery of misconfigured device; remove for production system

void pac_recovery(void)
{
    volatile uint32_t i;
    uint32_t perform_delay = 1;

    // PE3 is already an input, just need to Enable Pull Up
    PAC55XX_SCC->PEPUEN.P3 = 1;                         // Enable Pull Up

    // If PE3 is high a delay is not needed
    // Must check over the length of a full UART packet at 115kbps in case UART is talking and making PE3 low for short durations
    for (i=0; i < 40; i++)  // Value of 40 = at least 145 uS, which is enough for 1 UART packet (78uS)
    {
        // If PE3 is high at any time, then a delay is not needed
        if (PAC55XX_GPIOE->IN.P3 == 1)
        {
            perform_delay = 0;
        }
    }

    // If PE3 was low, delay required, add 6 second delay
    if(perform_delay)
    {
        pac_delay_asm(2800000);
    }

    // Restore PE3 to reset state
    PAC55XX_SCC->PEPUEN.P3 = 0;                         // Disable Pull Up
}

#else

void pac_recovery(void)
{
    // No Recovery required for production system
}

#endif  // #ifdef PAC_RECOVERY
